frontend/components/PaginatedList/PaginatedList.tsx

```diff
@@ -0,0 +1,189 @@
+  forwardRef,
+  Ref,
+} from "react";
+import { ReactElement } from "react-markdown/lib/react-markdown";
+import Checkbox from "components/forms/fields/Checkbox";
+import Spinner from "components/Spinner";
+// @ts-ignore
+import Pagination from "components/Pagination";
+
+const baseClass = "paginated-list";
+
+// Create an interface for the Ref, so that when parents use `useRef` to provide
+// a reference to this list, they can call `getDirtyItems()` on it to retrieve
+// the list of dirty items.
+export interface IPaginatedListHandle<TItem> {
+  getDirtyItems: () => TItem[];
+}
+interface IPaginatedListProps<TItem> {
+  // Function to fetch one page of data.
+  fetchPage: (pageNumber: number) => Promise<TItem[]>;
+  // UID property in an item. Defaults to `id`.
+  idKey?: string;
+  // Property to use as an item's label. Defaults to `name`.
+  labelKey?: string;
+  // How to determine whether to check an item's checkbox.
+  // If string, a key in an item whose truthiness will be checked.
+  // if function, a function that given an item, returns a boolean.
+  isSelected: string | ((item: TItem) => boolean);
+  // Custom function to render the label for an item.
+  renderItemLabel?: (item: TItem) => ReactElement | null;
+  // Custom function to render extra markup (besides the label) in an item row.
+  renderItemRow?: (
+    item: TItem,
+    // A callback function that the extra markup logic can call to indicate a change
+    // to the item, for example if a dropdown is changed.
+    onChange: (item: TItem) => void
+  ) => ReactElement | false | null | undefined;
+  // A function to call when an item's checkbox is toggled.
+  // Parents can use this to change whatever item metadata is needed to toggle
+  // the value indicated by `isSelected`.
+  onToggleItem: (item: TItem) => TItem;
+  // The size of the page to fetch and show.
+  pageSize?: number;
+  // The total # of items on all pages.
+  totalItems: number;
+}
+
+function PaginatedListInner<TItem extends Record<string, any>>(
+  {
+    fetchPage,
+    idKey: _idKey,
+    labelKey: _labelKey,
+    pageSize: _pageSize,
+    renderItemLabel,
+    renderItemRow,
+    onToggleItem,
+    isSelected,
+  }: IPaginatedListProps<TItem>,
+  ref: Ref<IPaginatedListHandle<TItem>>
+) {
+  // The # of the page to display.
+  const [currentPage, setCurrentPage] = useState(0);
+  // The set of items fetched via `fetchPage`.
+  const [items, setItems] = useState<TItem[]>([]);
+  // The set of items that have been changed in some way.
+  const [dirtyItems, setDirtyItems] = useState<Record<string | number, TItem>>(
+    {}
+  );
+  const [isLoading, setIsLoading] = useState(false);
+  const [error, setError] = useState<Error | null>(null);
+  const idKey = _idKey ?? "id";
+  const labelKey = _labelKey ?? "name";
+  const pageSize = _pageSize ?? 20;
+
+  // When the current page # changes, fetch a new page of data.
+  useEffect(() => {
+    let isCancelled = false;
+
+    async function loadPage() {
```

Comment:
Would it make sense to move this definition outside of the effect?

Reply:
Not in this case, because it's asynchronous, and `useEffect` callbacks can't be asynchronous.  Also it uses `isCancelled` as a closure variable.